### ISSUE TITLE
[机器学习波动率预测] | ML Volatility Prediction

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4392]
**标签**: quantitative-strategy, machine-learning, volatility-forecasting, deep-learning
### 摘要
使用LSTM、Transformer预测未来波动率，实现波动率择时交易。
### 算法描述
**核心原理**：
金融时间序列具有长程依赖性和非线性特征，传统GARCH模型难以捕捉复杂模式。LSTM通过门控机制选择性遗忘/记忆信息，Transformer通过自注意力机制捕捉长程依赖，联合预测精度显著提升。

**数学公式**：
- LSTM门控机制：
  - 遗忘门：$f_t = \sigma(W_f \cdot [h_{t-1}, x_t] + b_f)$
  - 输入门：$i_t = \sigma(W_i \cdot [h_{t-1}, x_t] + b_i)$
  - 候选记忆：$\tilde{C}_t = \tanh(W_C \cdot [h_{t-1}, x_t] + b_C)$
  - 细胞状态：$C_t = f_t \cdot C_{t-1} + i_t \cdot \tilde{C}_t$
  - 输出门：$o_t = \sigma(W_o \cdot [h_{t-1}, x_t] + b_o)$
  - 隐藏状态：$h_t = o_t \cdot \tanh(C_t)$

- Transformer自注意力：
  - $Attention(Q, K, V) = softmax(QK^T / \sqrt{d_k})V$
  - 多头注意力：$MultiHead(Q, K, V) = Concat(head_1, ..., head_h)W^O$

- 波动率预测：$\hat{\sigma}_{t+1} = f(X_t; \theta)$

**交易逻辑**：
1. 构建特征矩阵：历史收益率、成交量、技术指标、宏观经济数据
2. 训练LSTM/Transformer模型预测未来波动率
3. 当预测波动率↑，买入VIX期货或卖出期权
4. 波动率均值回归时反向平仓
### Python代码示例
```python
"""
机器学习波动率预测策略
使用LSTM、Transformer预测未来波动率，实现波动率择时交易
"""
import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
import warnings
warnings.filterwarnings('ignore')

# 模拟PyTorch接口
class MockTensor:
    def __init__(self, data):
        self.data = np.array(data)
    
    def numpy(self):
        return self.data

np.random.seed(42)

class MLVolatilityPrediction:
    """机器学习波动率预测策略"""
    
    def __init__(self, sequence_length: int = 60, prediction_horizon: int = 5,
                 model_type: str = "lstm"):
        self.sequence_length = sequence_length
        self.prediction_horizon = prediction_horizon
        self.model_type = model_type
        
        # 模型参数
        self.hidden_size = 64
        self.num_layers = 2
        self.num_heads = 4  # Transformer头数
        
        # 交易参数
        self.entry_threshold = 0.02
        self.position_size = 1.0
        
        # 训练状态
        self.is_trained = False
        
    def create_features(self, prices: pd.Series, volumes: pd.Series = None) -> np.ndarray:
        """
        构建特征矩阵
        特征包括：收益率、波动率、成交量变化、技术指标
        """
        features = []
        
        # 收益率特征
        returns = prices.pct_change().dropna()
        features.append(returns.values)
        
        # 历史波动率（滚动）
        for window in [5, 10, 20, 60]:
            roll_vol = returns.rolling(window).std().values
            features.append(roll_vol)
        
        # 收益率的滞后项
        for lag in range(1, 6):
            lagged = returns.shift(lag).values
            features.append(lagged)
        
        # 成交量变化
        if volumes is not None:
            vol_change = volumes.pct_change().fillna(0).values
            features.append(vol_change)
            vol_roll_std = volumes.rolling(20).std().fillna(0).values
            features.append(vol_roll_std)
        
        # 价格动量
        for period in [5, 10, 20]:
            momentum = prices.pct_change(period).values
            features.append(momentum)
        
        # 布林带位置
        for window in [20]:
            ma = prices.rolling(window).mean().values
            std = prices.rolling(window).std().values
            bb_position = (prices.values - ma) / (2 * std + 1e-8)
            features.append(bb_position)
        
        # 合并所有特征
        feature_matrix = np.column_stack(features)
        
        # 处理NaN
        feature_matrix = np.nan_to_num(feature_matrix, nan=0.0, posinf=0.0, neginf=0.0)
        
        return feature_matrix
    
    def normalize_features(self, X: np.ndarray, 
                          mean: np.ndarray = None, 
                          std: np.ndarray = None) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Z-score标准化"""
        if mean is None:
            mean = X.mean(axis=0)
        if std is None:
            std = X.std(axis=0) + 1e-8
            
        X_normalized = (X - mean) / std
        return X_normalized, mean, std
    
    def create_sequences(self, X: np.ndarray, y: np.ndarray = None) -> Tuple[np.ndarray, np.ndarray]:
        """创建时间序列样本"""
        sequences = []
        labels = []
        
        for i in range(self.sequence_length, len(X)):
            seq = X[i-self.sequence_length:i]
            sequences.append(seq)
            
            if y is not None:
                # 预测未来波动率
                label = y[i]  # 未来实际波动率
                labels.append(label)
        
        return np.array(sequences), np.array(labels) if labels else None
    
    def lstm_forward(self, X: np.ndarray, weights: Dict[str, np.ndarray]) -> np.ndarray:
        """
        LSTM前向传播
        X shape: (batch, seq_len, features)
        """
        batch_size, seq_len, features = X.shape
        
        # 初始化隐藏状态
        h = np.zeros((batch_size, self.hidden_size))
        C = np.zeros((batch_size, self.hidden_size))
        
        # 简化的LSTM计算（实际应用中应用PyTorch）
        for t in range(seq_len):
            x_t = X[:, t, :]
            
            # 遗忘门
            f_t = self._sigmoid(np.dot(x_t, weights['Wf']) + np.dot(h, weights['Uf']) + weights['bf'])
            # 输入门
            i_t = self._sigmoid(np.dot(x_t, weights['Wi']) + np.dot(h, weights['Ui']) + weights['bi'])
            # 候选记忆
            C_tilde = np.tanh(np.dot(x_t, weights['Wc']) + np.dot(h, weights['Uc']) + weights['bc'])
            # 输出门
            o_t = self._sigmoid(np.dot(x_t, weights['Wo']) + np.dot(h, weights['Uo']) + weights['bo'])
            
            # 更新状态
            C = f_t * C + i_t * C_tilde
            h = o_t * np.tanh(C)
        
        # 输出层
        output = np.dot(h, weights['Wy']) + weights['by']
        
        return output
    
    def transformer_encoder(self, X: np.ndarray, weights: Dict[str, np.ndarray]) -> np.ndarray:
        """
        Transformer编码器
        X shape: (batch, seq_len, features)
        """
        batch_size, seq_len, d_model = X.shape
        
        # 位置编码
        position_enc = self._get_positional_encoding(seq_len, d_model)
        X = X + position_enc
        
        # 多头自注意力
        Q = np.dot(X, weights['WQ'])
        K = np.dot(X, weights['WK'])
        V = np.dot(X, weights['WV'])
        
        # 缩放点积注意力
        d_k = d_model // self.num_heads
        Q = Q.reshape(batch_size, seq_len, self.num_heads, d_k).transpose(0, 2, 1, 3)
        K = K.reshape(batch_size, seq_len, self.num_heads, d_k).transpose(0, 2, 1, 3)
        V = V.reshape(batch_size, seq_len, self.num_heads, d_k).transpose(0, 2, 1, 3)
        
        attention = self._sigmoid(np.matmul(Q, K.transpose(0, 1, 3, 2)) / np.sqrt(d_k))
        attention = np.matmul(attention, V)
        
        # 合并多头
        attention = attention.transpose(0, 2, 1, 3).reshape(batch_size, seq_len, d_model)
        output = np.dot(attention, weights['WO']) + weights['bO']
        
        # 前馈网络
        ff_output = np.maximum(0, np.dot(output, weights['W1']) + weights['b1'])
        ff_output = np.dot(ff_output, weights['W2']) + weights['b2']
        
        # 池化
        pooled = np.mean(ff_output, axis=1)
        
        return pooled
    
    def _sigmoid(self, x: np.ndarray) -> np.ndarray:
        return 1 / (1 + np.exp(-np.clip(x, -500, 500)))
    
    def _get_positional_encoding(self, seq_len: int, d_model: int) -> np.ndarray:
        """位置编码"""
        position = np.arange(seq_len)[:, np.newaxis]
        div_term = np.exp(np.arange(0, d_model, 2) * -(np.log(10000) / d_model))
        
        pe = np.zeros((seq_len, d_model))
        pe[:, 0::2] = np.sin(position * div_term)
        pe[:, 1::2] = np.cos(position * div_term)
        
        return pe
    
    def initialize_weights(self, input_size: int) -> Dict[str, np.ndarray]:
        """初始化模型权重"""
        scale = 0.02
        hidden_size = self.hidden_size
        
        if self.model_type == "lstm":
            weights = {
                'Wf': np.random.randn(input_size, hidden_size) * scale,
                'Uf': np.random.randn(hidden_size, hidden_size) * scale,
                'bf': np.zeros(hidden_size),
                'Wi': np.random.randn(input_size, hidden_size) * scale,
                'Ui': np.random.randn(hidden_size, hidden_size) * scale,
                'bi': np.zeros(hidden_size),
                'Wc': np.random.randn(input_size, hidden_size) * scale,
                'Uc': np.random.randn(hidden_size, hidden_size) * scale,
                'bc': np.zeros(hidden_size),
                'Wo': np.random.randn(input_size, hidden_size) * scale,
                'Uo': np.random.randn(hidden_size, hidden_size) * scale,
                'bo': np.zeros(hidden_size),
                'Wy': np.random.randn(hidden_size, 1) * scale,
                'by': np.zeros(1)
            }
        else:  # transformer
            d_model = hidden_size
            weights = {
                'WQ': np.random.randn(d_model, d_model) * scale,
                'WK': np.random.randn(d_model, d_model) * scale,
                'WV': np.random.randn(d_model, d_model) * scale,
                'WO': np.random.randn(d_model, d_model) * scale,
                'bO': np.zeros(d_model),
                'W1': np.random.randn(d_model, d_model * 4) * scale,
                'b1': np.zeros(d_model * 4),
                'W2': np.random.randn(d_model * 4, d_model) * scale,
                'b2': np.zeros(d_model),
                'Wy': np.random.randn(d_model, 1) * scale,
                'by': np.zeros(1)
            }
        
        return weights
    
    def train(self, X_train: np.ndarray, y_train: np.ndarray,
             epochs: int = 50, batch_size: int = 32, learning_rate: float = 0.001):
        """训练模型"""
        # 标准化
        X_train_norm, self.mean, self.std = self.normalize_features(X_train)
        
        # 创建序列
        X_seq, y_seq = self.create_sequences(X_train_norm, y_train)
        
        # 初始化权重
        input_size = X_seq.shape[-1]
        self.weights = self.initialize_weights(input_size)
        
        # 简化训练循环（实际应用应用PyTorch）
        for epoch in range(epochs):
            # 随机梯度下降
            indices = np.random.permutation(len(X_seq))
            total_loss = 0
            
            for i in range(0, len(X_seq), batch_size):
                batch_idx = indices[i:i+batch_size]
                X_batch = X_seq[batch_idx]
                y_batch = y_seq[batch_idx].reshape(-1, 1)
                
                # 前向传播
                if self.model_type == "lstm":
                    pred = self.lstm_forward(X_batch, self.weights)
                else:
                    pred = self.transformer_encoder(X_batch, self.weights)
                    pred = np.dot(pred, self.weights['Wy']) + self.weights['by']
                
                # MSE损失
                loss = np.mean((pred - y_batch) ** 2)
                total_loss += loss
            
            if epoch % 10 == 0:
                print(f"Epoch {epoch}, Loss: {total_loss:.6f}")
        
        self.is_trained = True
        print(f"模型训练完成，最终损失: {total_loss:.6f}")
    
    def predict(self, X: np.ndarray) -> np.ndarray:
        """预测波动率"""
        if not self.is_trained:
            raise ValueError("模型未训练")
        
        X_norm = (X - self.mean) / self.std
        X_seq, _ = self.create_sequences(X_norm)
        
        if self.model_type == "lstm":
            pred = self.lstm_forward(X_seq, self.weights)
        else:
            pred = self.transformer_encoder(X_seq, self.weights)
            pred = np.dot(pred, self.weights['Wy']) + self.weights['by']
        
        return pred.flatten()
    
    def generate_signals(self, prices: pd.Series, volumes: pd.Series = None,
                        actual_volatility: pd.Series = None) -> pd.DataFrame:
        """
        生成交易信号
        返回: signals DataFrame
        """
        # 构建特征
        features = self.create_features(prices, volumes)
        
        signals = []
        
        for i in range(self.sequence_length + self.prediction_horizon, len(prices)):
            X = features[:i]
            y = actual_volatility.values[i-self.prediction_horizon:i] if actual_volatility is not None else None
            
            # 训练模型
            if not self.is_trained and y is not None:
                self.train(X, y)
            
            # 预测
            if self.is_trained:
                pred_vol = self.predict(X)[-1]
                current_vol = prices.pct_change().iloc[i-10:i].std()
                
                # 交易信号
                vol_change = pred_vol - current_vol
                
                if vol_change > self.entry_threshold:
                    signal = 1  # 买入波动率
                elif vol_change < -self.entry_threshold:
                    signal = -1  # 卖出波动率
                else:
                    signal = 0
                
                signals.append({
                    'date': prices.index[i],
                    'predicted_vol': pred_vol,
                    'current_vol': current_vol,
                    'signal': signal,
                    'confidence': abs(vol_change) / 0.05
                })
        
        return pd.DataFrame(signals)


# 示例使用
if __name__ == "__main__":
    strategy = MLVolatilityPrediction(
        sequence_length=60,
        prediction_horizon=5,
        model_type="lstm"
    )
    
    # 模拟价格数据
    dates = pd.date_range('2023-01-01', periods=500, freq='B')
    np.random.seed(42)
    
    # 生成模拟波动率数据
    returns = np.random.randn(500) * 0.02
    prices = 100 * np.exp(np.cumsum(returns))
    volumes = np.random.randint(1000000, 10000000, 500)
    
    prices = pd.Series(prices, index=dates)
    volumes = pd.Series(volumes, index=dates)
    
    # 计算实际波动率（用于训练）
    actual_vol = prices.pct_change().rolling(5).std() * np.sqrt(252)
    
    # 生成信号
    signals = strategy.generate_signals(prices, volumes, actual_vol)
    print(f"生成信号数量: {len(signals)}")
    print(signals.tail(20))
```
### 数据来源
- Reddit: r/algorithmictrading, r/MachineLearning, r/Finance
- 小红书: 量化投资、机器学习、深度学习量化
- 知乎: 量化交易、LSTM、Transformer应用
- 其他: arXiv (q-fin.LM), Papers With Code, DeepMind
